### ISSUE TITLE
Prohibiting two consecutive newlines at EOF does not work as expected

### ISSUE THREAD

**Issue by ghost** [eslint/eslint#12742]
**Tell us about your environment**

* **ESLint Version:** ^6.8.0
* **Node Version:** 10.0.0
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?** default (?)

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "env": {
    "es6": true,
    "node": true
  },
  "extends": [
    "airbnb-base"
  ],
  "globals": {
    "Atomics": "readonly",
    "SharedArrayBuffer": "readonly"
  },
  "parserOptions": {
    "ecmaVersion": 2018,
    "sourceType": "module"
  },
  "rules": {
    "no-multiple-empty-lines": ["error", { "max": 1 }]
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

Source code:

```js
// Source code of any .js file of the project
```

Command:

```bash
npx eslint .
```

**What did you expect to happen? What actually happened?**

I want to ensure there is *exactly one* newline at the end of each file. Since I am extending the Airbnb style guide, `eol-last` is already enabled, which afaik requires *at least one* newline at the end of the file. So all that's left for me to specify is that *more than one* newlines at the end of the file are disallowed. I thought I might be able to accomplish this like so:

```json
"no-multiple-empty-lines": ["error", { "max": 1 }]
```

Now, there should be both

* the requirement of *at least one* newline at the end of the file, due to `eol-last` being enabled in the Airbnb style guide, as well as
* the requirement of *no more than one* newline at a time (applying to the whole file, not just the end of the file), due to manually setting `no-multiple-empty-lines` to `{ "max": 1 }`.

However, this still configuration seems to still allow two consecutive newlines at the end of the file.

In the Gitter chat room, it was recommended to set `maxEOF` to `1`. This, however, does not change the fact that two newlines at the end of the file are being accepted. The fact that this does not fix the issue is to be expected, as `max` is already set to `1`, so setting `maxEOF` to `1` also should not have an effect.

In the Gitter chat room it was concluded to file a bug report.

**Are you willing to submit a pull request to fix this bug?**

Maybe. It would be an honor but I'm not sure I can. I'll take a look.

**Comment by kaicataldo** [comment]
I’m unable to reproduce this. I know you mentioned you did try the `maxEOF` option in the Gitter chat, but it looks like that is working correctly (see demo [here](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tbXVsdGlwbGUtZW1wdHktbGluZXNcIjogW1wiZXJyb3JcIiwgeyBcIm1heFwiOiAxLCBcIm1heEVPRlwiOiAxIH1dICovXG5cbmNvbnN0IGZvbyA9IDE7XG5cblxubGV0IGJhcjtcblxuXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7ImpzeCI6dHJ1ZX19LCJydWxlcyI6eyJhY2Nlc3Nvci1wYWlycyI6MiwiYXJyYXktYnJhY2tldC1uZXdsaW5lIjoyLCJhcnJheS1icmFja2V0LXNwYWNpbmciOjIsImFycmF5LWNhbGxiYWNrLXJldHVybiI6MiwiYXJyYXktZWxlbWVudC1uZXdsaW5lIjoyLCJhcnJvdy1ib2R5LXN0eWxlIjoyLCJhcnJvdy1wYXJlbnMiOjIsImFycm93LXNwYWNpbmciOjIsImJsb2NrLXNjb3BlZC12YXIiOjIsImJsb2NrLXNwYWNpbmciOjIsImJyYWNlLXN0eWxlIjoyLCJjYWxsYmFjay1yZXR1cm4iOjIsImNhbWVsY2FzZSI6MiwiY2FwaXRhbGl6ZWQtY29tbWVudHMiOjIsImNsYXNzLW1ldGhvZHMtdXNlLXRoaXMiOjIsImNvbW1hLWRhbmdsZSI6MiwiY29tbWEtc3BhY2luZyI6MiwiY29tbWEtc3R5bGUiOjIsImNvbXBsZXhpdHkiOjIsImNvbXB1dGVkLXByb3BlcnR5LXNwYWNpbmciOjIsImNvbnNpc3RlbnQtcmV0dXJuIjoyLCJjb25zaXN0ZW50LXRoaXMiOjIsImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJjdXJseSI6MiwiZGVmYXVsdC1jYXNlIjoyLCJkZWZhdWx0LXBhcmFtLWxhc3QiOjIsImRvdC1sb2NhdGlvbiI6MiwiZG90LW5vdGF0aW9uIjoyLCJlb2wtbGFzdCI6MiwiZXFlcWVxIjoyLCJmb3ItZGlyZWN0aW9uIjoyLCJmdW5jLWNhbGwtc3BhY2luZyI6MiwiZnVuYy1uYW1lLW1hdGNoaW5nIjoyLCJmdW5jLW5hbWVzIjoyLCJmdW5jLXN0eWxlIjoyLCJmdW5jdGlvbi1jYWxsLWFyZ3VtZW50LW5ld2xpbmUiOjIsImZ1bmN0aW9uLXBhcmVuLW5ld2xpbmUiOjIsImdlbmVyYXRvci1zdGFyLXNwYWNpbmciOjIsImdldHRlci1yZXR1cm4iOjIsImdsb2JhbC1yZXF1aXJlIjoyLCJncm91cGVkLWFjY2Vzc29yLXBhaXJzIjoyLCJndWFyZC1mb3ItaW4iOjIsImhhbmRsZS1jYWxsYmFjay1lcnIiOjIsImlkLWJsYWNrbGlzdCI6MiwiaWQtbGVuZ3RoIjoyLCJpZC1tYXRjaCI6MiwiaW1wbGljaXQtYXJyb3ctbGluZWJyZWFrIjoyLCJpbmRlbnQiOjIsImluZGVudC1sZWdhY3kiOjIsImluaXQtZGVjbGFyYXRpb25zIjoyLCJqc3gtcXVvdGVzIjoyLCJrZXktc3BhY2luZyI6Miwia2V5d29yZC1zcGFjaW5nIjoyLCJsaW5lLWNvbW1lbnQtcG9zaXRpb24iOjIsImxpbmVicmVhay1zdHlsZSI6MiwibGluZXMtYXJvdW5kLWNvbW1lbnQiOjIsImxpbmVzLWFyb3VuZC1kaXJlY3RpdmUiOjIsImxpbmVzLWJldHdlZW4tY2xhc3MtbWVtYmVycyI6MiwibWF4LWNsYXNzZXMtcGVyLWZpbGUiOjIsIm1heC1kZXB0aCI6MiwibWF4LWxlbiI6MiwibWF4LWxpbmVzIjoyLCJtYXgtbGluZXMtcGVyLWZ1bmN0aW9uIjoyLCJtYXgtbmVzdGVkLWNhbGxiYWNrcyI6MiwibWF4LXBhcmFtcyI6MiwibWF4LXN0YXRlbWVudHMiOjIsIm1heC1zdGF0ZW1lbnRzLXBlci1saW5lIjoyLCJtdWx0aWxpbmUtY29tbWVudC1zdHlsZSI6MiwibXVsdGlsaW5lLXRlcm5hcnkiOjIsIm5ldy1jYXAiOjIsIm5ldy1wYXJlbnMiOjIsIm5ld2xpbmUtYWZ0ZXItdmFyIjoyLCJuZXdsaW5lLWJlZm9yZS1yZXR1cm4iOjIsIm5ld2xpbmUtcGVyLWNoYWluZWQtY2FsbCI6Miwibm8tYWxlcnQiOjIsIm5vLWFycmF5LWNvbnN0cnVjdG9yIjoyLCJuby1hc3luYy1wcm9taXNlLWV4ZWN1dG9yIjoyLCJuby1hd2FpdC1pbi1sb29wIjoyLCJuby1iaXR3aXNlIjoyLCJuby1idWZmZXItY29uc3RydWN0b3IiOjIsIm5vLWNhbGxlciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNhdGNoLXNoYWRvdyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uZnVzaW5nLWFycm93IjoyLCJuby1jb25zb2xlIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29uc3RydWN0b3ItcmV0dXJuIjoyLCJuby1jb250aW51ZSI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWRpdi1yZWdleCI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZHVwbGljYXRlLWltcG9ydHMiOjIsIm5vLWVsc2UtcmV0dXJuIjoyLCJuby1lbXB0eSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1mdW5jdGlvbiI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXEtbnVsbCI6Miwibm8tZXZhbCI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRlbmQtbmF0aXZlIjoyLCJuby1leHRyYS1iaW5kIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLWxhYmVsIjoyLCJuby1leHRyYS1wYXJlbnMiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mbG9hdGluZy1kZWNpbWFsIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wbGljaXQtY29lcmNpb24iOjIsIm5vLWltcGxpY2l0LWdsb2JhbHMiOjIsIm5vLWltcGxpZWQtZXZhbCI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5saW5lLWNvbW1lbnRzIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pbnZhbGlkLXRoaXMiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1pdGVyYXRvciI6Miwibm8tbGFiZWwtdmFyIjoyLCJuby1sYWJlbHMiOjIsIm5vLWxvbmUtYmxvY2tzIjoyLCJuby1sb25lbHktaWYiOjIsIm5vLWxvb3AtZnVuYyI6Miwibm8tbWFnaWMtbnVtYmVycyI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVk).

I’d have to look at the Git history, but I’m assuming the `maxEOF` option exists so that this rule can be configured to not conflict with `eol-last`.

**Comment by jsn00b** [comment]
Like I described above: Setting `maxEOF` to `1` should not make a difference, since `max` is already set to `1`. In the docs it says:

> `maxEOF` can be used to set a different number for the end of file. The last blank lines will then be treated differently. **If omitted, the `max` option is applied at the end of the file.**

And it really **doesn't** make a difference, two consecutive lines at EOF are being permitted -- no matter whether `maxEOF` is explicitly set or not.

**Comment by mdjermanovic** [comment]
`eol-last` doesn't require an empty line at the end of the file. It requires that the last character in the source code is `\n`. There is no empty line at the end of the file after that last `\n`, although it looks like that in editors. You'd probably want to set `maxEOF` to `0`.

**Comment by jsn00b** [comment]
This certainly solved my issue. Thus, closing this issue is fine by me. However, you guys might want to leave it open, as this behavior is really non intuitive. Without the explanation that @mdjermanovic just gave, it's really hard to grasp why `maxEOF` has to be set to `0` to enforce exactly *one* line at the EOF.

**Comment by mdjermanovic** [comment]
> `maxEOF` has to be set to `0` to enforce exactly _one_ line at the EOF. Also, the documentation gives no clue about this.

I believe it's technically zero empty lines at the EOF, but it looks like one empty line in most editors.

I agree this should be explained in the documentation. Also, I think there should be an example with `maxEOF: 0` because it's a common configuration for this rule. PR is welcome!

Just, I'm not sure how to avoid this confusion in [maxEOF examples](https://eslint.org/docs/rules/no-multiple-empty-lines#maxeof) because the presentation depends on the editor/viewer.

For instance, if you open to view a source file on GitHub and see a code that looks exactly like the actual `incorrect` example in the docs (GitHub shows 8 lines, last 2 are empty), it's indeed an incorrect code by this rule. If you open the same file in VSCode, there will be one extra line (VSCode shows 9 lines, last 3 are empty).

However, if you open another source file in VSCode, and see a code that looks exactly like the actual `incorrect` example in the docs (VSCode shows 8 lines, last 2 are empty), it's actually a correct code by this rule.

![image](https://user-images.githubusercontent.com/44349756/71785607-d2535a80-3001-11ea-9337-b7258a20824d.png)

It's same in [ESLint Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tbXVsdGlwbGUtZW1wdHktbGluZXM6IFtcImVycm9yXCIsIHsgXCJtYXhcIjogMiwgXCJtYXhFT0ZcIjogMSB9XSovXG5cbnZhciBmb28gPSA1O1xuXG5cbnZhciBiYXIgPSAzO1xuXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjUsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

So, these examples can be indeed confusing. 

Also, it looks that some other rules like `max-lines` inconsistently treat this issue, because it seems that `SourceCode#lines` does contain an extra line after the last linebreak.

**Comment by kaicataldo** [comment]
Right, an empty line is actually `\n\n`. Would be great to clarify this in the documentation!

**Comment by mdjermanovic** [comment]
There is an option in VSCode to modify this behavior: `"editor.renderFinalNewline": false`

And it seems to be `true` by default on Windows only, `false` otherwise.

Also, the question does a newline/linebreak/eol sequence ends the current line or creates a new line (i.e. is there an empty line after the last `\n`) may be just a matter of convention in a particular environment, so it might be best to be neutral in the documentation - to just explain that editors/viewers/tools render and treat this differently and say that the rule doesn't count this as an empty line.

I was thinking to provide both views on the same examples in the docs  - one representing a screenshot of a viewer that doesn't render an extra empty line, the other representing a viewer that does render an extra empty line. (not real screenshots, just code blocks with extra lines)

Does this seem like a good idea?

There is also the following potential problem: code examples in the documentation have a different appearance on eslint.org and GitHub.

This looks like 2 empty lines at the end:

![image](https://user-images.githubusercontent.com/44349756/72480056-730dfb00-37f6-11ea-9409-a0685fe237c3.png)

This looks like 0 empty lines at the end:

![image](https://user-images.githubusercontent.com/44349756/72480015-54a7ff80-37f6-11ea-8fd0-ce5cb48fc1b6.png)

There are three `\n\n\n` on the site after `var bar = 3;`, while the GitHub viewer renders only one `\n`. 

This usually doesn't matter but in this particular case users might already get the wrong impression on how the rule works at the EOF depending on where they're reading the documentation.

**Comment by mdjermanovic** [comment]
It seems we all agree that the rule is working as intended and that it would be great to improve the documentation, so I marked this as an accepted "documentation" issue.

**Comment by kaicataldo** [comment]
> I was thinking to provide both views on the same examples in the docs - one representing a screenshot of a viewer that doesn't render an extra empty line, the other representing a viewer that does render an extra empty line. (not real screenshots, just code blocks with extra lines)

> Does this seem like a good idea?

Sounds good to me! I think it would helpful to show this with line numbers anyway.

**Comment by mdjermanovic** [comment]
> Sounds good to me! I think it would helpful to show this with line numbers anyway.

That's a great idea!

Is there already a way to show line numbers in code blocks (at least on eslint.org)?

**Comment by kaicataldo** [comment]
We don't have anything set up as far as I know, but it might be worth looking at the options for the Markdown -> HTML plugins we're using in Eleventy to see if there's a built-in way to do it. If not, we can implement it. That being said, I don't think that needs to be a blocker for the documentation update!
